kitsune/search/v2/search.py

```diff
@@ -0,0 +1,238 @@
+import bleach
+from datetime import datetime
+from abc import ABC, abstractmethod
+
+from elasticsearch_dsl import Search as DSLSearch, Q
+from kitsune.search.v2.es7_utils import es7_client
+from kitsune.search.config import WIKI_DOCUMENT_INDEX_NAME, QUESTION_INDEX_NAME
+from kitsune.wiki.config import CANNED_RESPONSES_CATEGORY, TEMPLATES_CATEGORY
+
+HIGHLIGHT_TAG = "strong"
+
+
+def first_highlight(hit):
+    highlight = getattr(hit.meta, "highlight", None)
+    if highlight:
+        return bleach.clean(
+            # `highlight` is of type AttrDict, which is internal to elasticsearch_dsl
+            # when converted to a dict, it's like:
+            # `{ 'es_field_name' : ['highlight1', 'highlight2'], 'field2': ... }`
+            # so here we're getting the first item in the first value in that dict:
+            next(iter(highlight.to_dict().values()))[0],
+            tags=[HIGHLIGHT_TAG],
+            strip=True,
+        )
+    return None
+
+
+def sanitize(summary):
+    return bleach.clean(
+        summary,
+        tags=[HIGHLIGHT_TAG],
+        strip=True,
+    )
+
+
+class BaseConfig(ABC):
+    """Semi-abstract base class for Search configs."""
+
+    @property
+    @abstractmethod
+    def index(self):
+        pass
+
+    @property
+    @abstractmethod
+    def fields(self):
+        """An array of fields to search over."""
+        pass
+
+    @property
+    @abstractmethod
+    def highlight_fields(self):
+        """An array of fields to highlight."""
+        pass
+
+    @property
+    def filter(self):
+        """A filter which returns all documents of this type to search over."""
+        return Q("term", _index=self.index)
+
+    @abstractmethod
+    def product_filter(self, product):
+        """A filter which returns documents with the correct product."""
+        pass
+
+    @abstractmethod
+    def make_result(self, hit, locale):
+        """Transform a hit into a result dictionary."""
+        pass
+
+
+class AAQConfig(BaseConfig):
+    """Search config for AAQ."""
+
+    index = QUESTION_INDEX_NAME
+    fields = [
+        "question_title.{}^4",
+        "question_content.{}^3",
+        "answer_content.{}^3",
+    ]
+    highlight_fields = ["question_content.{}", "answer_content.{}"]
+
+    @property
+    def filter(self):
+        return Q(
+            "bool",
+            filter=super().filter,
+            # exclude AnswerDocuments from the search:
+            must_not=Q("exists", field="updated"),
```

Comment:
can we use a type query here?

Reply:
Elastic itself doesn't know the types of QuestionDocuments and AnswerDocuments, just some have documents in question index have certain fields, whereas others don't.